crates/bevy_ecs/src/bundle.rs

```diff
@@ -1106,250 +1106,321 @@ impl<'w> BundleInserter<'w> {
         caller: MaybeLocation,
         relationship_insert_hook_mode: RelationshipInsertHookMode,
     ) -> (EntityLocation, T::Effect) {
-        let bundle_info = self.bundle_info.as_ref();
-        let archetype_after_insert = self.archetype_after_insert.as_ref();
-        let archetype = self.archetype.as_ref();
-
-        // SAFETY: All components in the bundle are guaranteed to exist in the World
-        // as they must be initialized before creating the BundleInfo.
-        unsafe {
-            // SAFETY: Mutable references do not alias and will be dropped after this block
-            let mut deferred_world = self.world.into_deferred();
+        unsafe fn before_write<'w, 'a>(
```

Comment:
It might help to leave a comment on each method explaining why it's broken up like this, so that some future dev doesn't try to put it back to improve legibility.